build_rules.gradle

```diff
@@ -478,65 +475,264 @@ ext.applyJavaNature = {
-              artifactId configuration.artifactId
-              // Strip the "shaded" classifier.
-              classifier null
-              // Set readable name to project description.
-              pom.withXml {
-                asNode().appendNode('name', description)
+    publishing {
+      repositories {
+        maven {
+          url(project.properties['distMgmtSnapshotsUrl'] ?: isRelease()
+                  ? 'https://repository.apache.org/service/local/staging/deploy/maven2'
+                  : 'https://repository.apache.org/content/repositories/snapshots')
+
+          // We attempt to find and load credentials from ~/.m2/settings.xml file that a user
+          // has configured with the Apache release and snapshot staging credentials.
+          // <settings>
+          //   <servers>
+          //     <server>
+          //       <id>apache.releases.https</id>
+          //       <username>USER_TOKEN</username>
+          //       <password>PASS_TOKEN</password>
+          //     </server>
+          //     <server>
+          //       <id>apache.snapshots.https</id>
+          //       <username>USER_TOKEN</username>
+          //       <password>PASS_TOKEN</password>
+          //     </server>
+          //   </servers>
+          // </settings>
+          def settingsXml = new File(System.getProperty('user.home'), '.m2/settings.xml')
+          if (settingsXml.exists()) {
+            def serverId = (isRelease() ? 'apache.releases.https' : 'apache.snapshots.https')
+            def m2SettingCreds = new XmlSlurper().parse(settingsXml).servers.server.find { server -> serverId.equals(server.id.text()) }
+            if (m2SettingCreds) {
+              credentials {
+                username m2SettingCreds.username.text()
+                password m2SettingCreds.password.text()
               }
             }
-            // TODO: Use the shadow test jar here instead of the test jar.
-            artifact packageTests {
-              classifier "tests"
+          }
+        }
+      }
+
+      publications {
+        mavenJava(MavenPublication) {
+          artifact shadowJar {
+            // Strip the "shaded" classifier.
+            classifier null
+          }
+          // TODO: Use the shadow test jar here instead of the test jar.
+          artifact packageTests {
+            classifier "tests"
+          }
+          artifact sourcesJar {
+            classifier "sources"
+          }
+          artifact testSourcesJar {
+            classifier "test-sources"
+          }
+          artifact javadocJar {
+            classifier "javadoc"
+          }
+
+          pom.withXml {
+            def root = asNode()
+            root.appendNode('name', project.description)
+            if (project.hasProperty("summary")) {
+                root.appendNode('description', project.summary)
             }
-            artifact sourcesJar {
-              classifier "sources"
+            root.appendNode('url', POM_URL)
```

Comment:
Do we get anything useful by making these gradle.properties?

It seems like each one is unique and we could just hardcode them all here.

Reply:
I saw this setup on another project doing similar maven publishing from Gradle.

1. Presumably it gives us the ability to access these variables from other places. Although I don't see that as a real use-case.
2. It removes some of the cruft from the build_rules.gradle file. But it would actually be more straightforward to have all the logic in one place.

Eventually I think it'd be cleaner to start with a template pom file with all of this filled in rather than build up the XML from code.

But for now, I'll move these settings into build_rules.gradle